### PR TITLE
Fix user earns while native Windows and macOS notifications are suppressed

### PR DESCRIPTION
**Pull request by tmancey**
Fixes https://github.com/brave/brave-browser/issues/3651

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [x] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [x] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

- Confirm native Ad notifications are not shown on Windows if Focus Assist is enabled
- Confirm native Ad notifications are not shown on Windows if notifications are disabled for the user
- Confirm native Ad notifications are not shown on Windows if notifications are disabled for application
- Confirm native Ad notifications are not shown on Windows if notifications are disabled by group policy
- Confirm native Ad notifications are not shown on Windows if notifications are disabled by manifest
- Confirm native Ad notifications are not shown on macOS 10.15 if notifications set to none
- Confirm native Ad notifications are not shown on macOS 10.14 and older if notifications set to none
- Confirm ads are shown if native notifications are set to banner or alert

Note: Notifications will always be shown for Chromium Message Center (Brave fallback if native notifications are not supported)

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

**Comment by tmancey** [comment]
@bbondy @bridiver Would be great if you could code review as to detect if Focus Assist is enabled I am having to use undocumented code, thanks

File: components/brave_ads/browser/notification_helper_win.cc [hunk 1]
```diff
@@ -27,4 +113,129 @@ NotificationHelper* NotificationHelper::GetInstance() {
   return NotificationHelperWin::GetInstance();
 }
 
+///////////////////////////////////////////////////////////////////////////////
+
+bool NotificationHelperWin::IsFocusAssistEnabled() const {
+  const auto nt_query_wnf_stated_data_func =
```
Comment (bridiver) [inline components/brave_ads/browser/notification_helper_win.cc]: doesn't this need a null check?

File: components/brave_ads/browser/notification_helper_win.cc [hunk 2]
```diff
@@ -27,4 +113,129 @@ NotificationHelper* NotificationHelper::GetInstance() {
...
+      GetProcAddress(GetModuleHandle(L"ntdll"), "NtQueryWnfStateData");
+
+  const auto nt_query_wnf_stated_data_ptr =
```
Comment (bridiver) [inline components/brave_ads/browser/notification_helper_win.cc]: I believe this should be `const auto*`, but not sure why the google checks in the compiler aren't complaining about it

File: components/brave_ads/browser/notification_helper_win.cc (hunk 1 above)
Comment (tmancey) [reply components/brave_ads/browser/notification_helper_win.cc]: @bridiver This does not require a null check as the next call fails if this is null

File: components/brave_ads/browser/notification_helper_win.cc (hunk 2 above)
Comment (tmancey) [reply components/brave_ads/browser/notification_helper_win.cc]: I have refactored the code and removed ptr as changing to const auto * caused a compiler error

File: components/brave_ads/browser/ads_service_impl.cc [hunk 3]
```diff
@@ -1922,7 +1923,7 @@ void AdsServiceImpl::ShowNotification(
 #endif
 }
 
-bool AdsServiceImpl::ShouldShowNotifications() const {
+bool AdsServiceImpl::ShouldShowNotifications() {
```
Comment (masparrow) [inline components/brave_ads/browser/ads_service_impl.cc]: de-const'ing?

Comment (masparrow) [reply components/brave_ads/browser/ads_service_impl.cc]: and all the other de-const's - presuming a valid reason but curious...

File: components/brave_ads/browser/notification_helper_linux.cc [hunk 4]
```diff
@@ -13,11 +12,19 @@ NotificationHelperLinux::NotificationHelperLinux() = default;
 
 NotificationHelperLinux::~NotificationHelperLinux() = default;
 
-bool NotificationHelperLinux::ShouldShowNotifications() const {
-  return !IsFullScreenMode();
+bool NotificationHelperLinux::ShouldShowNotifications() {
+  // TODO(tmancey): Investigate how we can detect if notifications are enabled
+  // within the Linux operating system
```
Comment (masparrow) [inline components/brave_ads/browser/notification_helper_linux.cc]: Checking the policy with TODO's like this - would this be better as a new ticket?

File: components/brave_ads/browser/notification_helper_mac.mm [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        case UNAuthorizationStatusProvisional: {
+          is_authorized = true;
+          break;
+        }
+      }
+
+      dispatch_semaphore_signal(semaphore);
+    }];
+
+    dispatch_semaphore_wait(semaphore, DISPATCH_TIME_FOREVER);
```
Comment (masparrow) [inline components/brave_ads/browser/notification_helper_mac.mm]: I get nervous about 'FOREVER'... are we _that_ committed to blocking this thread forever if OSX FUBAR's?

File: components/brave_ads/browser/ads_service_impl.cc (hunk 3 above)
Comment (tmancey) [reply components/brave_ads/browser/ads_service_impl.cc]: Unfortunately, some of the calls could no longer be const due to changing of state

File: components/brave_ads/browser/notification_helper_mac.mm (hunk 5 above)
Comment (tmancey) [reply components/brave_ads/browser/notification_helper_mac.mm]: If macOS foobar's we have bigger problems, and ads are running in their own service

File: components/brave_ads/browser/notification_helper_linux.cc (hunk 4 above)
Comment (tmancey) [reply components/brave_ads/browser/notification_helper_linux.cc]: TODO's are allowed as long as a ticket has been raised, however good point so I have updated the TODO to contain a link to the issue

**Pull request merged**